basiclti/basiclti-blis/src/java/org/sakaiproject/lti13/LTI13Servlet.java

```diff
@@ -295,6 +301,60 @@ protected void doPost(HttpServletRequest request, HttpServletResponse response)
 
 	}
 
+	// A very locked down proxy - JSON Only
+	protected void handleProxy(HttpServletRequest request, HttpServletResponse response) {
+		String proxyUrl = request.getParameter("proxyUrl");
+		if ( proxyUrl == null ) {
+			LTI13Util.return400(response, "Missing proxyUrl");
+			return;
+		}
+
+		Session sess = SessionManager.getCurrentSession();
+		if ( sess == null || sess.getUserId() == null ) {
+			LTI13Util.return400(response, "Must be logged in");
+			return;
+		}
+
+		try {
+			java.net.URL url = new java.net.URL(proxyUrl);
+			java.net.HttpURLConnection con = (java.net.HttpURLConnection) url.openConnection();
+			con.setRequestMethod("GET");
+			con.setConnectTimeout(3000);
+			con.setReadTimeout(3000);
+			con.setInstanceFollowRedirects(true);
+
+			java.io.BufferedReader in = new java.io.BufferedReader(
+				new java.io.InputStreamReader(con.getInputStream()));
+			String inputLine;
+			StringBuffer content = new StringBuffer();
+			while ((inputLine = in.readLine()) != null) {
+				content.append(inputLine);
+			}
+			if ( content.length() > 10000 ) {
+				in.close();
+				LTI13Util.return400(response, "Content too long");
+				return;
+			}
+			in.close();
```

Comment:
better to close() in a finally block

Reply:
OK - Thanks.

Reply:
I actually disagree.  What if the exception happens before the "in" is defined (i.e. in the first 6 lines of the try block) - how would I handle that in a finally clause?

Reply:
You can't even reference "in" in the finally clause unless I hoist it outside the try block.  Which is super icky practice IMHO.

Reply:
You can use a {{try}} with resources, then you can close the reader in a finally block:

https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html

Reply:
Yes - but I make two things in the try block - both of which might fail - I could (a) hoist all that into the try or (b) outside of the try completely or (c) make two try blocks.  I think the way I have it is by far the clearest expression of what is happening - but if you both insist I think that it should be two try blocks with hoists.  But to be honest - it makes really easy to read code - almost unreadable.

Reply:
idk about code readability here. whenever i see connections opened without a definitive way of closing it (try-with-resources or finally), i get code smell....

assuming your code is going to close the conn, or return the conn, or close the SecurityAdvisor as expected seems like a bad recipe

Reply:
I agree with Sam here, this is standard practise nowadays.

Reply:
This is a **crap** pattern.

I will not be rewriting this just to make use of the cool new parenthesis feature.  I tried a rewrite that looked far uglier - and it turns out that thing go bad if there is an exception in the () part - What the ever loving F!  The simple example @bjones86 references of a con.createStatement() - it is a a good but pointless use of the pattern.  Because it will always work and never throws any exceptions and its close() can be called if the follow on code fails.  It is like a pointless tautology dreamed up to prove the point.  The previous example with Zip files - must have exceptions happen if things go wrong in the parenthesis code.  If the bit in the parenthesis does not allow proper handling of exceptions - then it is of supremely limited value.

Feel free to "prove me wrong" by sending me a "better version" of the code.

Reply:
```
try (HttpURLConnection con = new URL(...); BufferReader in = new BufferedReader(...))
{ ... }
catch (...) { ... }
finally
{
    in.close();
}
```

Reply:
I will try it - but what if the URL is a 404 and the () bit blows up - my initial testing is that nothing is caught and it just blows up.  Am I wrong with that?

Reply:
[ERROR] /Users/csev/dev/sakai-scripts/trunk/basiclti/basiclti-blis/src/java/org/sakaiproject/lti13/LTI13Servlet.java:[319,38] incompatible types: try-with-resources not applicable to variable type
[ERROR]     (java.net.URL cannot be converted to java.lang.AutoCloseable)
[ERROR] /Users/csev/dev/sakai-scripts/trunk/basiclti/basiclti-blis/src/java/org/sakaiproject/lti13/LTI13Servlet.java:[320,52] incompatible types: try-with-resources not applicable to variable type
[ERROR]     (java.net.HttpURLConnection cannot be converted to java.lang.AutoCloseable)

Reply:
yeah i think you do the HttpUrlConnection first and then try-with-resources on the BufferedReader

https://stackoverflow.com/questions/43708021/proper-use-of-urlconnection-and-try-with-resources-in-java-8

Reply:
Yeah - That is a try within a try.  That can work. Ugly but not wrong.

Reply:
I think I got it!  I was being confused by my code logging the exception in my catch block - I kept thinking the catch block was being ignored.